src/lib/drivers/device/Device.hpp

```diff
@@ -114,6 +114,23 @@ class __EXPORT Device
 	 */
 	virtual int	write(unsigned address, void *data, unsigned count) { return -ENODEV; }
 
+	/**
+	 * Read a register from the device.
+	 *
+	 * @param		The register to read.
+	 * @return		The value that was read.
+	 */
+	virtual uint8_t read_reg(unsigned reg) { return -ENODEV; }
```

Comment:
There might be multiple variations for 8/16/... bits. You also need a way to check if a transfer failed.
```suggestion
	virtual int read_reg8(unsigned reg) { return -ENODEV; }
```

Reply:
There are already named as read_reg and read_reg16 in existing code. The  return type was done so much earlier on. This pull request initially only armed to get 2 I2C drivers and options for bus mode and frequency in, now the code change scope is much bigger than what's expected initially. I have no way to test the change even if I make it, so it's better to have one more experienced person with these driver code to make the systematic changes.

Reply:
Yes exactly, which is why I suggested not to change anything at the `Device.hpp` level, as I don't want to burden you having to go through all drivers (of course you're welcome to do more cleanup).
What you can do to reduce scope is like in BMP280, adding a common base: https://github.com/PX4/Firmware/blob/master/src/drivers/barometer/bmp280/bmp280.h#L133.

Reply:
With so many sensor drivers re-implement read_reg and write_reg API, I think my code change is one step ahead in correct direction. For anyone else who is going to add new SPI/I2C drivers like what I need do this time, he/she will benefit from my change which sets up abstract APIs at base class and provides an implementation for POSIX platform. I did not remove existing implementations (mostly duplicated ones)  for other sensors which I have no way to test, but others could do so earlier than doing it from scratch. So may we keep my change and improve read_reg/write_reg step by step when developers get chance and hardware to work on them and verify further changes? The wide usage of read_reg/write_reg API indicates the need of them at higher level in the initial phase.

Reply:
There might be a bigger refactoring in this area, but for now I'm fine with this.

Reply:
> There might be a bigger refactoring in this area, but for now I'm fine with this.

Likewise. Looking forward to refactoring all this nearly duplicate code out of existence.